concordium-std/src/test_infrastructure.rs

```diff
@@ -1460,9 +1478,46 @@ impl<State: Serial + DeserialWithState<TestStateApi> + StateClone<TestStateApi>>
         if amount.micro_ccd > 0 {
             *self.contract_balance.borrow_mut() -= amount;
         }
+        // Update the receiver query balance.
+        if let Some(balance) = self.query_contract_balances.borrow_mut().get_mut(to) {
+            *balance += amount;
+        }
+
         Ok(res)
     }
 
+    fn account_balance(&self, address: AccountAddress) -> QueryAccountBalanceResult {
+        if self.missing_accounts.contains(&address) {
+            Err(QueryAccountBalanceError)
+        } else if let Some(balances) = self.query_account_balances.borrow().get(&address) {
+            Ok(*balances)
+        } else {
+            fail!("No account balance for {:?} has been set up.", address)
+        }
+    }
+
+    fn contract_balance(&self, address: ContractAddress) -> QueryContractBalanceResult {
+        // If the contract address is set and matches, we return the contract balance
+        // instead.
+        if Some(address) == self.contract_address {
+            Ok(self.contract_balance.borrow().to_owned())
```

Comment:
I am wondering whether it is necessay to have this special case. What goes wrong if we handle it generically?

Reply:
Ok, we are going to move away from this sort of testing anyhow. 

This does have rather awkward semantics to explain though so I am not convinced that keeping backwards compatibility is enough of a reason. But at this time I don't htink it is worth changing the approach either.